diff --git a/target/linux/mediatek/Makefile b/target/linux/mediatek/Makefile
index c8ab5e0..01e993d 100644
--- a/target/linux/mediatek/Makefile
+++ b/target/linux/mediatek/Makefile
@@ -5,7 +5,7 @@ include $(TOPDIR)/rules.mk
 ARCH:=arm
 BOARD:=mediatek
 BOARDNAME:=MediaTek Ralink ARM
-SUBTARGETS:=mt7622 mt7623 mt7629
+SUBTARGETS:=mt7622 mt7623 mt7629 mt7986 mt7981
 FEATURES:=squashfs nand ramdisk fpu
 
 KERNEL_PATCHVER:=5.4
